### PR TITLE
Loans: principal and interest separation

### LINKED ISSUE #1379: Loans: principal & interest separation
### Description

Currently, we repay an amount, but it should internally be treated separately as the principal amount and the interest amount. [slack thread](https://kflabs.slack.com/archives/C050E3ZN2JU/p1685448669053569)

### PR DESCRIPTION
**Pull request by lemunozm**
# Description

Epic [specs](https://centrifuge.hackmd.io/@jeroen/HyZrkvxDh)
Slack [channel](https://centrifugedao.slack.com/archives/C05DX0RH1T3)

Fixes #1379

## Features/enhancements
- Repayments done with principal and interest amounts
- External pricing now calculates interests

## Changes and Descriptions

- [x] Repay separately principal and interest
  - [x] For internal pricing
  - [x] For external pricing 
- [x] Write-off penalty for external pricing: We should probably create a new entity that stores: `interest_rate`, `normalized_debt`, and `write_off_penalty` properties to be used in both internal and external pricing and abstract ours from the interest rate properties.
- [x] Allow interest mutation in external pricing 
- [x] Update tests
- [x] Update benchmarks

### CODE REVIEW

File: pallets/loans/src/entities/loans.rs
Comment (offerijns) [inline pallets/loans/src/entities/loans.rs:327]: What if `ExternalActivePricing.outstanding_amount()` was rewritten to return separate outstanding principal and interest? Something like

```rust
#[derive(Default, Encode, Decode, Clone, PartialEq, Eq, TypeInfo, RuntimeDebug, MaxEncodedLen)]
pub struct OutstandingAmount<Balance> {
  principal: Balance,
  interest: Balance,
}

impl OutstandingAmount<Balance> {
	pub fn total(&self) -> Result<Balance, ArithmeticError> {
		self.principal().ensure_add(self.interest())
	}
}

impl<T: Config> ExternalActivePricing<T> {
        ...

	pub fn outstanding_amount(&self) -> Result<OutstandingAmount, ArithmeticError> {
		let interest_reference_amount = self
			.info
			.notional
			.ensure_mul_int(self.outstanding_quantity)?;

		OutstandingAmount {
			principal: self.total_borrowed.ensure_sub(self.total_repaid.principal)?,
			interest: Ok(self.calculate_debt()?.ensure_sub(interest_reference_amount)?)
		}
	}

```

File: pallets/loans/src/entities/pricing/external.rs
Comment (offerijns) [inline pallets/loans/src/entities/pricing/external.rs:92]: I would only use the term principal for the "to be repaid amount excluding interest", so rather here something like

```suggestion
		let outstanding_notional = self
```

File: pallets/loans/src/types/mod.rs
Comment (offerijns) [inline pallets/loans/src/types/mod.rs:203]: This is a nice abstraction!

File: pallets/loans/src/entities/loans.rs
Comment (lemunozm) [reply pallets/loans/src/entities/loans.rs:325]: Ok, I see the idea. Maybe by now, because it's only used in one place and it's not stored, could we use a tuple instead?

**Comment by lemunozm** [comment]
Last architecture after adding the `ActiveInterestRate`:
image

File: pallets/loans/src/entities/interest.rs
Comment (offerijns) [inline pallets/loans/src/entities/interest.rs:75]: Would probably be good to consistently use the `base interest rate + penalty = interest rate` terminology
```suggestion
	pub fn set_base_interest_rate(&mut self, base_interest_rate: T::Rate) -> DispatchResult {
```

Comment (lemunozm) [reply pallets/loans/src/entities/interest.rs]: Make sense!

File: pallets/loans/src/entities/loans.rs
Comment (offerijns) [reply pallets/loans/src/entities/loans.rs:325]: > Ok, I see the idea. Maybe by now, because it's only used in one place and it's not stored nowwhere, could we use a tuple instead?

Yeah that's fine 👍 

> In the example, principal will be interest_reference_amount, right? Because total_borrowed and total_repaid are not accessible from ExternalActivePricing

Ah yeah I didn't consider that it would not be accessible from there. Never mind then

**Review by offerijns — DISMISSED** [review]
I think it would be good to get a second review on this (maybe @mustermeiszer or @branan ), but the business logic changes look good to me!

File: pallets/loans/src/entities/pricing/external.rs
Comment (mustermeiszer) [reply pallets/loans/src/entities/pricing/external.rs:144]: I am trying to understand what we are doing here. 

---
* any letter or number with an elvated annotation of $DIV$, eg. $x^{DIV}$, should be a number in the decimals of `T::Rate`
* $DIV$ is just $10^{decimals}$
---
What we have here given we want to adjust the principal by `100`. Assuming pool currency has NO decimals.
* $p$ is the price
* quantitiy: $q = \frac{100 * DIV}{p^{DIV}} * DIV$

And we are checking at the end in the ensure clause:
$p^{DIV} * q = 100$ &rarr; $p^{DIV} * \frac{100 * DIV}{p^{DIV}} * DIV = 100$

where `ensure_mul_int` is NOT adapting the decimals of `100`.

Comment (mustermeiszer) [reply pallets/loans/src/entities/pricing/external.rs:144]: I am either not writing this down correctly (probably) or something is off here ^^

File: pallets/loans/src/entities/pricing/external.rs
Comment (mustermeiszer) [inline pallets/loans/src/entities/pricing/external.rs:156]: In case we are reducing by too much this errors out, right?

File: pallets/loans/src/entities/pricing/external.rs
Comment (mustermeiszer) [inline pallets/loans/src/entities/pricing/external.rs:163]: Why must the notinal be multiplied by the `quantity`? Is the notional always the value of `1` asset? 

Could you maybe explain that a bit? I.e. how the interest adjustment works?

File: pallets/loans/src/entities/loans.rs
Comment (mustermeiszer) [inline pallets/loans/src/entities/loans.rs:351]: Can we change the `RepayRestrictions` for an active loan? I yes, this `is_zero()` check will prevent repaying the loan. I also think it is not adding any value or what should the problem be, with having a non zero value here?

**Review by mustermeiszer — COMMENTED** [review]
Overall looks really good. I have a few deeper quetsions, which most likely come from me not understanding the logic correctly. But I would like to clarify those before approving.

File: pallets/loans/src/entities/pricing/external.rs
Comment (offerijns) [reply pallets/loans/src/entities/pricing/external.rs:163]: >  Is the notional always the value of 1 asset?

Yeah exactly, notional value is the value of 1 asset (e.g. 1 T-bill). This can basically be considered the "reference amount that interest accrues on". This notional value never changes, only the price (valuation) does.

So the interest should accrue on `notional x outstanding quantity`

File: pallets/loans/src/entities/interest.rs
Comment (lemunozm) [reply pallets/loans/src/entities/interest.rs:20]: Sure, I forgot to document it!

File: pallets/loans/src/entities/loans.rs
Comment (lemunozm) [reply pallets/loans/src/entities/loans.rs:64]: It already was in the `InternalPricing` entity. The check has been moved here because now it's common to all pricing methods.

File: pallets/loans/src/entities/loans.rs
Comment (lemunozm) [reply pallets/loans/src/entities/loans.rs]: Added docs. Probably the method has mutated enough to have a better name, maybe `prepare_repayment(amount)`?

File: pallets/loans/src/entities/pricing/external.rs
Comment (lemunozm) [reply pallets/loans/src/entities/pricing/external.rs:144]: Let me check this deeply 👍🏻

File: pallets/loans/src/entities/pricing/external.rs
Comment (lemunozm) [reply pallets/loans/src/entities/pricing/external.rs:156]: Yes, with an `ArithmeticError::Underflow`.

File: pallets/loans/src/entities/pricing/external.rs
Comment (lemunozm) [reply pallets/loans/src/entities/pricing/external.rs:163]: The notional is the price used as a reference to calculate interest. Here's the [slack thread](https://centrifugedao.slack.com/archives/C05DX0RH1T3/p1687857589371469) with the @offerijns explanation about this:

File: pallets/loans/src/entities/loans.rs
Comment (lemunozm) [reply pallets/loans/src/entities/loans.rs]: We can not change repay restrictions right now (not sure if we should in the future).

`FullOnce` is, in fact, two restrictions (that in the future will be split: #1353). We want to repay the entire amount, and we want to repay it once. 

But thinking more on it, I think, as you say, checking `is_zero()` makes no sense. And even could generate inconsistency. i.e. You borrow again, but you can not repay again because you repaid the first time. Probably the correct restriction is just `Full`. Good catch!

File: pallets/loans/src/entities/pricing/external.rs
Comment (mustermeiszer) [reply pallets/loans/src/entities/pricing/external.rs:163]: But why do we need the multiplication? Shouldn't the `interest` simply be added?

File: pallets/loans/src/entities/pricing/external.rs
Comment (lemunozm) [reply pallets/loans/src/entities/pricing/external.rs:144]: Well, first, some questions/considerations about the types:
- Currently, `Price` is a `Rate`, but should it be? what should be a `Balance` and what should be a `Rate`? still not clear to me.
- Quantity is a `Balance`, but it's just a multiplier to that price, maybe it deserves another type.

Given said that and supposing 2-decimal precision, if my price is `2000.01` and my quantity is 5, my `Balance` as a result of multiplying this should be `1000005`, right?

---

> Will the `principa_adj` already have the right decimals?

Whatever `principal` is, comes from multiplying the current price by quantity on the client side. So it should have the right decimals if the client side did it well (we should probably create an RPC for this). i.e. if the price is `2000.01` and quantity 5, it should be `10000.05` that interpreted as balance is 1000005

> I think this should rather be `T::Rate::from_inner()`

Still need to think more about this, but I think you're right. 
Testing is passing, but I think I'm making a bad assumption calculating the principle as follows: 
```rust		
let amount = PRICE_VALUE.saturating_mul_int(QUANTITY);
```
which makes me lose the price decimals.

> Given we do not want to saturate here, or do we want to?

I think we should check it, yes. We do not want to saturate here 👍🏻 

> * quantitiy: $q = \frac{100 * DIV}{p^{DIV}} * DIV$

Why not just $q = \frac{100 * DIV}{p^{DIV}}$ without the last `DIV`? from where does the last `DIV` come?

> where `ensure_mul_int` is NOT adapting the decimals of 100

Not sure what you mean here. `principal` is a `Balance`, so any "decimal" is part of the integer value itself that it is multiplying by the rate.

Comment (lemunozm) [reply pallets/loans/src/entities/pricing/external.rs:144]: I do not know if I have enough information from the abstract type to do the correct maths.

What I need is an integer `Balance` (which I do not know the precision convention) as a result of a multiplication of a `Rate` by a multiplier. How do I know how many decimals from the price I should keep for the `Balance` result?

File: pallets/loans/src/entities/pricing/external.rs
Comment (lemunozm) [reply pallets/loans/src/entities/pricing/external.rs:163]: Because the notional is the price of only one treasury, you can have `quantity` treasuries

File: pallets/loans/src/entities/pricing/external.rs
Comment (lemunozm) [reply pallets/loans/src/entities/pricing/external.rs:144]: When would the u128 actually overflow?

I'm going to check it. 

Anyways I think from the `pallet-loans` perspective, I have no way to be sure if the user is using a `Rate` with more precision than a `Balance`. I think the big issue here is that we are mixing `Balance` with `Rate` to represent the "same" thing. Maybe we need to choose how to represent any kind of currency value in `Balance` or `Rate` and not mix them... 🤔

File: pallets/loans/src/entities/interest.rs
Comment (offerijns) [inline pallets/loans/src/entities/interest.rs:66]: `RateCollection.current_debt` expects an interest rate per sec, while this is passing an interest rate per year. How does that work? 🤔

Comment (lemunozm) [reply pallets/loans/src/entities/interest.rs:66]: It actually expects an interest rate per year from #1221. It makes the conversion inside

File: pallets/loans/src/entities/pricing/external.rs
Comment (branan) [reply pallets/loans/src/entities/pricing/external.rs:144]: We've definitely ended up using `Rate` where we shouldn't. This has generally been done where we want to keep working in the 27-decimal fixed-point space rather than keeping integers around (I don't believe we have a different balance type).

Ideally, we'd be using `Balance` anywhere we represent a quantity, and only use Rate for, well, rates. Interest accrual, penalties, etc should all be done in the `Rate` space, and then multiplied by the integer `Balance`

In this case, since `quantity` is also a Balance (though maybe we should make that a different integer type, since it doesn't represent a currency amount?), I think we don't need any specific fancy fixed-point math, or even checked math:

```
quantity = principal / price;
ensure!(principal % price == 0)
```

Comment (lemunozm) [reply pallets/loans/src/entities/pricing/external.rs:144]: Thanks @branan, for coming to the conversation. I think you've pointed out the origin of the issue correctly. 

Actually, `Price` was a `Balance` initially, but we changed after seeing other chains using rates for that in #1387 (but if I remember well @offerijns, we had both no strong opinion on doing it, just seemed more "appropriated"). Seeing now how a `Rate` price behaves now with the rest of the amount values makes much more sense to change it again to `Balance` and avoid all these issues.

Comment (lemunozm) [reply pallets/loans/src/entities/pricing/external.rs:144]: > since quantity is also a Balance (though maybe we should make that a different integer type, since it doesn't represent a currency amount?)

Yes, it doesn't represent a currency amount, only a multiplier. However, you could see this as if this quantity is, in fact, the number of "tokens" of a T-bill currency.

Comment (offerijns) [reply pallets/loans/src/entities/pricing/external.rs:144]: For quantity, I do think we want some decimal precision ideally (so you can have a quantity of 0.1). So would probably keep that as a Balance.

For changing price to a Balance, what would the decimals of that then be though? Right now, the decimals of all Balances are always based on the pool currency, while Rate is a fixed 27 decimals. The prices can be used by multiple pools so cannot be linked to 1 pool currency. I still don't really see why changing price to a Balance would solve the above issue, but I might not fully follow it.

Comment (lemunozm) [reply pallets/loans/src/entities/pricing/external.rs:144]: If we need to compute the result as Balance, we do not know how many decimals from the Rate I should choose.

Nevertheless, having price as a Balance, no matter the precision the user chooses, it will be the same for both the price and the amount. The oracle should be responsable from setting a correct price with the required Balance format.

Comment (lemunozm) [reply pallets/loans/src/entities/pricing/external.rs:144]: Now I think, that Quantity should be a Rate, because it acts as a multiplier of another type.

Comment (lemunozm) [reply pallets/loans/src/entities/pricing/external.rs:144]: > Right now, the decimals of all Balances are always based on the pool currency

Could you expand this @offerijns? Are we using different Balances with different decimal precision?

Comment (offerijns) [reply pallets/loans/src/entities/pricing/external.rs:144]: Yes all `Balance` properties in a pool (e.g. all invested amounts, the reserve, ...) as well as all the loan properties (e.g. the collateral value, borrowed amount, ...) are based on the decimal precision of the pool currency.

**Comment by lemunozm** [comment]
I think this can be merged and leave the issue for another PR

**Review by mustermeiszer — APPROVED** [review]
Approving! 

NOTE: We need to solve #1447 before deploying this to production.

File: pallets/loans/src/entities/pricing/external.rs
Comment (lemunozm) [reply pallets/loans/src/entities/pricing/external.rs:144]: Will be here: https://github.com/centrifuge/centrifuge-chain/issues/1447

**Pull request merged**
